File: android/TourGraph/app/src/main/java/com/nikhilsi/tourgraph/model/Tour.kt
package com.nikhilsi.tourgraph.model

import android.database.Cursor
import java.util.Locale

data class Tour(
    val id: Int,
    val productCode: String,
    val title: String,
    val oneLiner: String?,
    val description: String?,
    val destinationId: Int,
    val destinationName: String?,
    val country: String?,
    val continent: String?,
    val timezone: String?,
    val rating: Double?,
    val reviewCount: Int?,
    val fromPrice: Double?,
    val currency: String?,
    val durationMinutes: Int?,
    val imageUrl: String?,
    val imageUrlsJson: String?,
    val highlightsJson: String?,
    val inclusionsJson: String?,
    val viatorUrl: String?,
    val supplierName: String?,
    val weightCategory: String?
) {
    val displayPrice: String
        get() {
            val price = fromPrice ?: return ""
            val symbol = when (currency) {
                "USD" -> "$"
                "EUR" -> "€"
                "GBP" -> "£"
                else -> currency ?: "$"
            }
            return if (price == price.toLong().toDouble()) {
                "$symbol${price.toLong()}"
            } else {
                "$symbol${String.format(Locale.US, "%.2f", price)}"
            }
        }

    val displayRating: String
        get() {
            val r = rating ?: return ""
            return String.format(Locale.US, "%.1f", r)
        }

    val displayDuration: String
        get() {
            val mins = durationMinutes ?: return ""
            return when {
                mins >= 1440 -> {
                    val days = mins / 1440
                    val remainHours = (mins % 1440) / 60
                    if (remainHours > 0) "${days}d ${remainHours}h" else "${days}d"
                }
                mins >= 60 -> {
                    val hours = mins / 60
                    val remainMins = mins % 60
                    if (remainMins > 0) "${hours}h ${remainMins}m" else "${hours}h"
                }
                else -> "${mins}m"
            }
        }

    val imageURL: String? get() = imageUrl

    val highlights: List<String>
        get() {
            val json = highlightsJson ?: return emptyList()
            return try {
                kotlinx.serialization.json.Json.decodeFromString<List<String>>(json)
            } catch (_: Exception) {
                emptyList()
            }
        }

    val imageURLs: List<String>
        get() {
            val json = imageUrlsJson ?: return emptyList()
            return try {
                kotlinx.serialization.json.Json.decodeFromString<List<String>>(json)
            } catch (_: Exception) {
                emptyList()
            }
        }

    companion object {
        val cardColumns = arrayOf(
            "id", "product_code", "title", "one_liner", "description",
            "destination_id", "destination_name", "country", "continent", "timezone",
            "rating", "review_count", "from_price", "currency", "duration_minutes",
            "image_url", "image_urls_json", "highlights_json", "inclusions_json",
            "viator_url", "supplier_name", "weight_category"
        )

        fun fromCursor(cursor: Cursor): Tour {
            return Tour(
                id = cursor.getInt(cursor.getColumnIndexOrThrow("id")),
                productCode = cursor.getString(cursor.getColumnIndexOrThrow("product_code")) ?: "",
                title = cursor.getString(cursor.getColumnIndexOrThrow("title")) ?: "",
                oneLiner = cursor.getStringOrNull("one_liner"),
                description = cursor.getStringOrNull("description"),
                destinationId = cursor.getInt(cursor.getColumnIndexOrThrow("destination_id")),
                destinationName = cursor.getStringOrNull("destination_name"),
                country = cursor.getStringOrNull("country"),
                continent = cursor.getStringOrNull("continent"),
                timezone = cursor.getStringOrNull("timezone"),
                rating = cursor.getDoubleOrNull("rating"),
                reviewCount = cursor.getIntOrNull("review_count"),
                fromPrice = cursor.getDoubleOrNull("from_price"),
                currency = cursor.getStringOrNull("currency"),
                durationMinutes = cursor.getIntOrNull("duration_minutes"),
                imageUrl = cursor.getStringOrNull("image_url"),
                imageUrlsJson = cursor.getStringOrNull("image_urls_json"),
                highlightsJson = cursor.getStringOrNull("highlights_json"),
                inclusionsJson = cursor.getStringOrNull("inclusions_json"),
                viatorUrl = cursor.getStringOrNull("viator_url"),
                supplierName = cursor.getStringOrNull("supplier_name"),
                weightCategory = cursor.getStringOrNull("weight_category")
            )
        }

        private fun Cursor.getStringOrNull(column: String): String? {
            val idx = getColumnIndex(column)
            return if (idx >= 0 && !isNull(idx)) getString(idx) else null
        }

        private fun Cursor.getDoubleOrNull(column: String): Double? {
            val idx = getColumnIndex(column)
            return if (idx >= 0 && !isNull(idx)) getDouble(idx) else null
        }

        private fun Cursor.getIntOrNull(column: String): Int? {
            val idx = getColumnIndex(column)
            return if (idx >= 0 && !isNull(idx)) getInt(idx) else null
        }
    }
}

File: ios/TourGraph/TourGraph/Shared/Tour.swift
import Foundation
import GRDB

/// Core tour model — maps directly to the `tours` SQLite table.
struct Tour: Identifiable, Codable, FetchableRecord, Sendable {
    let id: Int
    let productCode: String
    let title: String
    let oneLiner: String?
    let description: String?
    let destinationId: String?
    let destinationName: String?
    let country: String?
    let continent: String?
    let timezone: String?
    let latitude: Double?
    let longitude: Double?
    let rating: Double?
    let reviewCount: Int?
    let fromPrice: Double?
    let currency: String?
    let durationMinutes: Int?
    let imageUrl: String?
    let imageUrlsJson: String?
    let highlightsJson: String?
    let inclusionsJson: String?
    let viatorUrl: String?
    let supplierName: String?
    let weightCategory: String?

    // MARK: - Column mapping (snake_case DB → camelCase Swift)

    enum CodingKeys: String, CodingKey, ColumnExpression {
        case id
        case productCode = "product_code"
        case title
        case oneLiner = "one_liner"
        case description
        case destinationId = "destination_id"
        case destinationName = "destination_name"
        case country
        case continent
        case timezone
        case latitude
        case longitude
        case rating
        case reviewCount = "review_count"
        case fromPrice = "from_price"
        case currency
        case durationMinutes = "duration_minutes"
        case imageUrl = "image_url"
        case imageUrlsJson = "image_urls_json"
        case highlightsJson = "highlights_json"
        case inclusionsJson = "inclusions_json"
        case viatorUrl = "viator_url"
        case supplierName = "supplier_name"
        case weightCategory = "weight_category"
    }

    // MARK: - Computed helpers

    var displayPrice: String {
        guard let price = fromPrice else { return "" }
        if price >= 1000 {
            return "$\(Int(price).formatted())"
        }
        return "$\(Int(price))"
    }

    var displayRating: String {
        guard let r = rating else { return "" }
        return String(format: "%.1f", r)
    }

    var displayDuration: String {
        guard let mins = durationMinutes else { return "" }
        if mins >= 1440 {
            let days = mins / 1440
            let remainingHrs = (mins % 1440) / 60
            if remainingHrs > 0 {
                return "\(days)d \(remainingHrs)h"
            }
            return "\(days) day\(days == 1 ? "" : "s")"
        } else if mins >= 60 {
            let hrs = mins / 60
            let remainingMins = mins % 60
            if remainingMins > 0 {
                return "\(hrs)h \(remainingMins)m"
            }
            return "\(hrs) hr\(hrs == 1 ? "" : "s")"
        }
        return "\(mins) min"
    }

    var imageURL: URL? {
        guard let str = imageUrl else { return nil }
        return URL(string: str)
    }

    var viatorURL: URL? {
        guard let str = viatorUrl else { return nil }
        return URL(string: str)
    }

    var highlights: [String] {
        guard let json = highlightsJson,
              let data = json.data(using: .utf8),
              let arr = try? JSONDecoder().decode([String].self, from: data)
        else { return [] }
        return arr
    }

    var imageURLs: [URL] {
        guard let json = imageUrlsJson,
              let data = json.data(using: .utf8),
              let arr = try? JSONDecoder().decode([String].self, from: data)
        else { return [] }
        return arr.compactMap { URL(string: $0) }
    }

    /// Columns needed for card display (roulette hand, superlatives, etc.)
    static let cardColumns: [CodingKeys] = [
        .id, .productCode, .title, .oneLiner, .destinationName, .country,
        .continent, .rating, .reviewCount, .fromPrice, .currency,
        .durationMinutes, .imageUrl, .viatorUrl, .weightCategory
    ]
}
